src/adapter/src/catalog/state.rs

```diff
@@ -2263,6 +2267,34 @@ impl CatalogState {
             SystemObjectId::System => SystemObjectType::System,
         }
     }
+
+    /// Synchronizes the local view of the [`StorageMetadata`] with the
+    /// [`Transaction`]'s.
+    ///
+    /// This must be called after any `Transaction` is given to the storage
+    /// controller, otherwise subsequent storage operations will have
+    /// inconsistent metadata.
+    pub(super) fn update_storage_metadata(&mut self, tx: &Transaction<'_>) {
+        use mz_storage_client::controller::StorageTxn;
+        self.storage_metadata.collection_metadata = tx.get_collection_metadata();
```

Comment:
This is just a nit so its not required, but the best option I think is:

```
pub(super) with_storage_metadata<F, R>(&mut self, tx: &Transaction<'_>, sc: Option<&mut dyn StorageController>, callback: F) -> R
where
    F: FnOnce(&mut StorageController, &Transaction<'_>) -> R {
    if let Some(sc) = sc {
        callback(sc, tx);
		self.updated_storage_metadata(tx);
    }
}
```

Reply:
storage_controller calls are async tho, so this is probably not worth it

Reply:
Ah, I tried to do something like this and gave up because it was "saucy." Let me give it another go and will report back if I can make it not-gnarly.

Reply:
This winds up being the implementation:


```rust
    pub(super) async fn with_storage_metadata<F, Fut>(
        &mut self,
        tx: &mut Transaction<'_>,
        sc: &mut dyn StorageController<Timestamp = mz_repr::Timestamp>,
        f: F,
    ) -> Result<(), StorageError<mz_repr::Timestamp>>
    where
        F: FnOnce(
            &mut dyn StorageController<Timestamp = mz_repr::Timestamp>,
            &mut Transaction<'_>,
        ) -> Fut,
        Fut: Future<Output = Result<(), StorageError<mz_repr::Timestamp>>>,
    {
        let r = f(sc, tx).await;
        if r.is_ok() {
            self.update_storage_metadata(tx);
        }
        r
    }
```

I think you and I can read this just fine, but I fear the opacity it introduces for folks who are less experienced with Rust. Gonna leave this here in case we find a use for it later but don't think we should merge it. On the upside, if you miss a call to `update_storage_metadata`, you'll realize it because the word melts around you (ask how I know :))